# Assignment 3 Submission


- Name: Kenji Otsuka
- Email: kotsuka@umich.edu
- GitHub Repo: [https://github.com/KenjiOhtsuka/UoM_SIADS521_Assignment3](https://github.com/KenjiOhtsuka/UoM_SIADS521_Assignment3)
- Publicly Available Dashboard:
- Video Link:

## Index

- [Data explanation](#data-explanation)
- [Visualization Technique](#visualization-technique)
- [Visualization Library](#visualization-library)
- [Explorative Data Analysis](#explorative-data-analysis)
- [Demonstration](#demonstration)

<a name="data-explanation"></a>
## Data explanation

In this assignment, I used the Gapminder dataset, which contains data on life expectancy, GDP per capita, and population for countries around the world from 1952 to 2007.
The dataset is well-suited for exploring long-term trends in health and economic development.

- **Dataset**: Gapminder: Countries Over Time
- **Author**: SHRADDHA, Brian Risk
- **Source**: https://www.kaggle.com/datasets/shraddha4ever20/gapminder-countries-over-time
- **License**: Creative Commons Attribution 4.0 International (CC BY 4.0)

Note. The original source of this dataset is not explicitly stated on Kaggle. It may include modifications made by the uploader.


<a name="visualization-technique"></a>
## Visualization Technique

### Description of each visualization type

### How the visualizations complement each other and when each should be used.

In the demonstration we use **line charts**, **scatter plots**, **violin plots** including bar plots and jittered points, **bar charts** and **histograms**. Bar charts are used only in Explorative Data Analysis in advance of the demonstration. Each visualization serves a different purpose, and together they provide a complete view of long‑term trends, regional differences, and relationships among variables.

**Line charts** are used to show how indicators change over time.
They are effective for identifying long‑term trends, comparing continents, and observing periods of rapid improvement or decline.

**Scatter plots** reveal relationships between variables with actual data points.
Scatter plots also use color and point size to represent continent and population, helping users compare categories at a glance.

**Violin plots**, combined with jittered points and box plots, summarize the distribution of a variable within each category.
Jittered points show individual points, box plots highlight the median and interquartile range, and the violin shape shows the overall density. This combination provides a detailed view of variation within each category.

**Bar charts** are helpful when the goal is to compare magnitudes rather than trends or relationships.

**Histograms** show the distribution of a variable. They help users understand how values are spread globally and whether the distribution is narrow, wide, or skewed.

### How the visualizations are combined into a dashboard and how they complement each other.

The dashboard contains Line charts, Histograms, and Scatter plots.
Each visualization highlights a different aspect of the Gapminder dataset, and together they provide a clear and interactive view of global health and economic development.

**Line charts** show how life expectancy changes over time for the selected country.
A vertical red line marks the chosen year, helping users connect long‑term trends with the specific year shown in the other charts.

**Histograms** display the global distribution of life expectancy in the selected year.
The bin corresponding to the selected country is highlighted in red, making it easy to see where that country falls within the worldwide distribution.

**Scatter plots** show the relationship between GDP per capita and life expectancy for the selected year.
The chosen country is emphasized with a red star marker, allowing users to compare it directly with other countries.

These visualizations complement each other:

- **Line charts** show **trends over time**,
- **Histograms** show **global distributions**,
- **Scatter plots** show **relationships**.

With ipywidgets, all charts update simultaneously when the user changes the country or year.
This coordinated interactivity allows users to explore long‑term trends, compare countries, and understand global patterns within a single dashboard.

<a name="visualization-library"></a>
## Visualization Library

I have chosen to use Plotly for creating interactive visualizations, and ipywidgets for adding interactivity to the dashboard.

### Plotly

Plotly is an open-source graphing library developed by Plotly Technologies Inc., with contributions from the open-source community. It allows for the creation of interactive and visually appealing plots. It supports a wide range of chart types, including scatter plots, line charts, bar charts, and more. Additionally, Plotly integrates well with Jupyter notebooks, making it easy to create and display visualizations within the notebook environment.

Plotly follows a hybrid declarative and imperative design. Users can specify visual attributes declaratively through figure objects while fine-grained layout adjustments and callbacks are handled imperatively. This makes Plotly flexible for both simple and advanced visualizations.

The limitations include a steeper learning curve for detailed customization and potential performance issues with very large datasets. However, its interactivity and customization options make it a powerful tool for data visualization.

It is suitable for this assignment because it provides rich interactivity, it can be used in Jupyter Notebooks, it can create multiple types of visualizations.

- Installation: Plotly can be installed using pip with the command `pip install plotly`.
- URL: [https://plotly.com/python/](https://plotly.com/python/)

### Ipywidgets

Ipywidgets is a library that provides interactive widgets for Jupyter notebooks. It is a part of the Jupyter ecosystem and is maintained by the Jupyter community with contributions from the open-source community. It allows users to create sliders, dropdowns, buttons, and other interactive elements that can be used to control visualizations and data analysis workflows.

Ipywidgets is a procedural library. It means that users create and manipulate widgets through code. This approach allows for fine-grained control over the behavior and appearance of widgets.

The limitations include limited styling options and potential compatibility issues with certain Jupyter notebook versions. However, it is a valuable tool for adding interactivity to visualizations and enhancing the user experience.

It is suitable for this assignment because it allows real-time updates to Plotly visualizations and enables multi-visualization interactive dashboards.

- Installation: Ipywidgets can be installed using pip with the command `pip install ipywidgets`.
- URL: [https://ipywidgets.readthedocs.io/en/stable/](https://ipywidgets.readthedocs.io/en/stable/)

<a name="explorative-data-analysis"></a>
## Explorative Data Analysis

Here, I looked basic attributes of the dataset.

In [ ]:
import pandas as pd

if os.path.exists("gap.csv"):
    df = pd.read_csv("gap.csv")
else:
    df = pd.read_csv("https://raw.githubusercontent.com/KenjiOhtsuka/UoM_SIADS521_Assignment3/refs/heads/main/gap.csv")

df.head()

,country,continent,year,lifeExp,pop,gdpPercap,iso_alpha,iso_num
0,Afghanistan,Asia,1952,28.801,8425333,779.445314,AFG,4
1,Afghanistan,Asia,1957,30.332,9240934,820.853030,AFG,4
2,Afghanistan,Asia,1962,31.997,10267083,853.100710,AFG,4
3,Afghanistan,Asia,1967,34.020,11537966,836.197138,AFG,4
4,Afghanistan,Asia,1972,36.088,13079460,739.981106,AFG,4


The dataset contains the following columns:

| column name | description |
|-------------|-------------|
| country     | name of the country |
| continent   | name of the continent |
| year        | year of measurement |
| lifeExp     | life expectancy at birth (in years) |
| pop         | total population |
| gdpPercap   | GDP per capita (in USD) |
| iso_alpha   | ISO 3166-1 alpha country code |
| iso_num     | ISO 3166-1 numeric country code |

The number of categories and the range of values are:

In [2]:
print("The number of countries:", df["country"].nunique())
print("The number of continents:", df["continent"].nunique())
print("The number of years:", df["year"].nunique(), "(", df["year"].min(), "-", df["year"].max(), ")")

# summary statistics of the dataset
df.describe()

The number of countries: 142
The number of continents: 5
The number of years: 12 ( 1952 - 2007 )


,year,lifeExp,pop,gdpPercap,iso_num
count,1704.00000,1704.000000,1.704000e+03,1704.000000,1704.000000
mean,1979.50000,59.474439,2.960121e+07,7215.327081,425.880282
std,17.26533,12.917107,1.061579e+08,9857.454543,248.305709
min,1952.00000,23.599000,6.001100e+04,241.165876,4.000000
25%,1965.75000,48.198000,2.793664e+06,1202.060309,208.000000
50%,1979.50000,60.712500,7.023596e+06,3531.846989,410.000000
75%,1993.25000,70.845500,1.958522e+07,9325.462346,638.000000
max,2007.00000,82.603000,1.318683e+09,113523.132900,894.000000


The dataset contains data for 142 countries across 5 continents.

"Year" ranges from 1952 to 2007, but the number of unique values are 12 not 2007 - 1952 + 1 = 56. 
`value_counts()` shows the detail. 

In [3]:
df['year'].value_counts()

year
1952    142
1957    142
1962    142
1967    142
1972    142
1977    142
1982    142
1987    142
1992    142
1997    142
2002    142
2007    142
Name: count, dtype: int64

We can see the dataset contains data from 1952 to 2007, with measurements taken every 5 years.

In [4]:
df = df.sort_values(by=['continent', 'country', 'year']).reset_index(drop=True)

### Correlation of numeric variables

In [69]:
import plotly.express as px

fig = px.imshow(
    df.select_dtypes(include=['float64', 'int64']).drop(columns=['iso_num']).corr(),
    text_auto=True,
    color_continuous_scale='RdBu',
    aspect="auto"
)
fig.update_layout(autosize=False, width=800, height=500)
fig.show()

In [6]:
df.select_dtypes(include=['float64', 'int64']).drop(columns=['iso_num']).corr()

,year,lifeExp,pop,gdpPercap
year,1.000000,0.435611,0.082308,0.227318
lifeExp,0.435611,1.000000,0.064955,0.583706
pop,0.082308,0.064955,1.000000,-0.025600
gdpPercap,0.227318,0.583706,-0.025600,1.000000


Correlation shows life expectancy and GDP per capita is moderately and positively correlated (0.58).

In [7]:
df["gdp"] = df["gdpPercap"] * df["pop"]

### Life expectancy over time

According to the correlation analysis, life expectency and year are moderately and positively correlated (0.43).
We can see the trend in line chart below.

In [8]:
import plotly.express as px
import plotly.graph_objects as go

base_colors = px.colors.qualitative.Plotly
continents = df["continent"].unique()
continent_colors = {
    cont: base_colors[i % len(base_colors)]
    for i, cont in enumerate(continents)
}

def hex_to_rgba(hex_color, alpha=0.3):
    hex_color = hex_color.lstrip('#')
    r = int(hex_color[0:2], 16)
    g = int(hex_color[2:4], 16)
    b = int(hex_color[4:6], 16)
    return f"rgba({r},{g},{b},{alpha})"


In [9]:

fig = go.Figure()

for country, d in df.groupby("country"):
    cont = d["continent"].iloc[0]
    base_color = continent_colors[cont]
    rgba_color = hex_to_rgba(base_color, alpha=0.3)

    fig.add_trace(go.Scatter(
        x=d["year"],
        y=d["lifeExp"],
        mode="lines",
        line=dict(color=rgba_color, width=2),
        name=country,
        legendgroup=cont,
        showlegend=False
    ))

# continent ごとの凡例を追加
for cont, base_color in continent_colors.items():
    fig.add_trace(go.Scatter(
        x=[None], y=[None],
        mode="lines",
        line=dict(color=base_color, width=3),
        name=cont
    ))

fig.update_layout(
    title="Life Expectancy Over Time",
    xaxis_title="Year",
    yaxis_title="Life Expectancy"
)

fig.show()


It shows the life expectency has been increasing over time for most of the countries, but we can see decreasing trend for some countries and years. It seems that life expectency in Oceania and European countries is higher and one in African countries lower. 

Which countries have the highest life expectency increase between 1952 and 2007?

In [10]:
df_growth_1952_2007 = df[df["year"].isin([1952, 2007])].pivot(index="country", columns="year", values="lifeExp").reset_index()
df_growth_1952_2007["growth_rate"] = df_growth_1952_2007[2007] - df_growth_1952_2007[1952]
df_growth_1952_2007.sort_values("growth_rate", ascending=False).head(10)

year,country,1952,2007,growth_rate
96,Oman,37.578,75.640,38.062
137,Vietnam,40.412,74.249,33.837
59,Indonesia,37.468,70.650,33.182
109,Saudi Arabia,39.875,72.777,32.902
75,Libya,42.723,73.952,31.229
70,"Korea, Rep.",47.453,78.623,31.170
92,Nicaragua,42.314,72.899,30.585
138,West Bank and Gaza,43.160,73.422,30.262
139,"Yemen, Rep.",32.548,62.698,30.150
46,Gambia,30.000,59.448,29.448


Then, take a look at mean life expectancy by continent over time.

In [11]:
df_continent = df.groupby(["continent", "year"]).agg({"lifeExp": ["mean", "std"], "gdp": ["sum"], "pop": "sum"}).reset_index()
df_continent.columns = [
    f"{col[0]}_{col[1]}" if isinstance(col, tuple) and col[1] else col[0]
    for col in df_continent.columns
]
df_continent["gdp_percap"] = df.groupby(["continent", "year"]).apply(lambda x: x["gdp"].sum() / x["pop"].sum()).values
df_continent.head()


,continent,year,lifeExp_mean,lifeExp_std,gdp_sum,pop_sum,gdp_percap
0,Africa,1952,39.135500,5.151581,3.115993e+11,237640501,1311.221439
1,Africa,1957,41.266346,5.620123,3.826778e+11,264837738,1444.951993
2,Africa,1962,43.319442,5.875364,4.568136e+11,296516865,1540.599054
3,Africa,1967,45.334538,6.082673,5.950877e+11,335289489,1774.847446
4,Africa,1972,47.450942,6.416258,7.837566e+11,379879541,2063.171343


In [46]:
fig = px.line(
    df_continent,
    x="year",
    y="lifeExp_mean",
    color="continent",
    title="Average Life Expectancy by Continent",
    labels={"continent": "Continent", "lifeExp_mean": "Average Life Expectancy"}
)
fig.update_layout(autosize=False, width=800, height=500)
fig.show()


It shows that mean of life expectency by continent has been increasing over time except for Africa, and the order of mean life expectency by continent is Oceania > Europe > Americas > Asia > Africa for all years.

### GDP by capita over time

Now, let's look at GDP per capita over time by continent.

In [47]:
fig = px.line(
    df_continent,
    x="year",
    y="gdp_percap",
    color="continent",
    title="Average GDP per Capita by Continent"
)
fig.update_layout(autosize=False, width=800, height=500)
fig.show()
    

The order of mean GDP per capita by continent is Oceania > Europe > Americas/Asia > Africa. In 1980s, Asia and America order was reversed, but it has been reversed again since 1990s. This order is almost the same as the order of mean life expectency by continent.

In [14]:
### Population by continent over time

In [48]:
fig = px.line(
    df_continent,
    x="year",
    y="pop_sum",
    color="continent",
    title="Total Population by Continent"
)
fig.update_layout(autosize=False, width=800, height=500)
fig.show()


Total population by continent in Asia remarkably increased over time, but Asian population is too large and it minimizes the population of other continents. So, I use log scale and index to show the population change over time.

In [49]:
# population in log scale
fig = px.line(
    df_continent,
    x="year",
    y="pop_sum",
    color="continent",
    title="Total Population by Continent (Log Scale)",
    log_y=True
)
fig.update_layout(autosize=False, width=800, height=500)
fig.show()

Log scale just shows the difference clearly.

In [17]:
df_continent["pop_index"] = df_continent["pop_sum"] / df_continent.groupby("continent")["pop_sum"].transform("first") * 100
df["pop_index"] = df["pop"] / df.groupby("country")["pop"].transform("first") * 100


In [50]:
fig = px.line(
    df,
    x="year",
    y="pop_index",
    color="continent",
    title="Population Index by Country",
    log_y=True
)
fig.update_layout(autosize=False, width=800, height=500)
fig.show()

In [51]:

fig = px.line(
    df_continent,
    x="year",
    y="pop_index",
    color="continent",
    title="Total Population by Continent"
)
fig.update_layout(autosize=False, width=800, height=500)
fig.show()


Index chart shows another aspect of population change over time. The lowest index is Europe and one of the highest is Africa..

### Relation among GDP per capita, life expectancy, and population

In [52]:
fig = px.scatter(
    df,
    x="gdpPercap",
    y="lifeExp",
    color="continent",
    size="pop",
    hover_name="country",
    title="GDP per Capita vs Life Expectancy",
    labels={"gdpPercap": "GDP per Capita", "lifeExp": "Life Expectancy"},
    log_x=True,
)
fig.update_layout(autosize=False, width=800, height=500)
fig.show()


In [53]:
fig = px.scatter(
    df,
    x="pop",
    y="gdpPercap",
    color="continent",
    hover_name="country",
    title="Population vs GDP per Capita",
    labels={"pop": "Population", "gdpPercap": "GDP per Capita"},
    log_x=True,
    log_y=True
)
fig.update_layout(autosize=False, width=800, height=500)
fig.show()


In [54]:
fig = px.scatter(
    df_continent[df_continent["pop_sum"] > 0],
    x="pop_sum",
    y="gdp_percap",
    color="continent",
    # size=df_continent["pop_sum"] * 100,
    title="Continent-Level GDP vs Life Expectancy",
    labels={"gdp_percap": "Average GDP per Capita", "lifeExp_mean": "Average Life Expectancy"},
    log_x=True,
    log_y=True,
)
import math
# fig.update_xaxes(range=[
#     7,
#     # math.log10(df_continent["pop_sum"].min() * 0.9),
#     math.log10(df_continent["pop_sum"].max() * 1.5)
# ])
# # fig.update_yaxes(range=[
# #     math.log10(df_continent["gdpPercap_mean"].min() * 0.9),
# #     math.log10(df_continent["gdpPercap_mean"].max() * 1.1)
# # ])
fig.update_layout(autosize=False, width=800, height=500)
fig.show()

In [55]:
fig = px.scatter(
    df,
    x="pop",
    y="lifeExp",
    color="continent",
    hover_name="country",
    title="Population vs Life Expectancy",
    labels={"pop": "Population", "lifeExp": "Life Expectancy"},
    opacity=0.7,
    log_x=True,
)
fig.update_layout(autosize=False, width=800, height=500)
fig.show()


In [56]:
fig = px.scatter(
    df_continent,
    x="gdp_percap",
    y="lifeExp_mean",
    color="continent",
    # size=(df_continent["pop_sum"] * 1000),
    title="Continent-Level GDP vs Life Expectancy",
    labels={"gdp_percap": "Average GDP per Capita", "lifeExp_mean": "Average Life Expectancy"},
    log_x=True,
)
fig.update_layout(autosize=False, width=800, height=500)
fig.show()


### Global life expectancy inequality over time

Standard deviation of life expectency by country over time shows the global life expectency inequality has been decreasing over time.
It means the gap of life expectency between countries.

In [57]:
df_std = df.groupby("year")["lifeExp"].std().reset_index()

fig = px.line(
    df_std,
    x="year",
    y="lifeExp",
    title="Global Life Expectancy Inequality Over Time (Std Dev)",
    labels={"lifeExp": "Life Expectancy Standard Deviation"}
)
fig.update_layout(autosize=False, width=800, height=500)
fig.show()


Global life expectancy inequality is not in constant decrease. It decreases to 1987 but increases later.

In [26]:
df_continent["gdp_growth"] = df_continent.groupby("continent")["gdp_percap"].pct_change()


In [58]:
fig = px.line(
    df_continent,
    x="year",
    y="gdp_growth",
    color="continent",
    title="GDP Growth Rate by Continent"
)
fig.update_layout(autosize=False, width=800, height=500)
fig.show()


In [59]:
df_growth = df.groupby("country")["lifeExp"].pct_change().reset_index()
df_growth = pd.concat([df["country"], df["year"], df_growth["lifeExp"]], axis=1)

fig = px.imshow(
    df_growth.pivot(index="country", columns="year", values="lifeExp"),
    aspect="auto",
    color_continuous_scale="RdBu",
    title="Life Expectancy Growth Rate Heatmap"
)
fig.update_layout(autosize=False, width=800, height=500)
fig.show()


In [ ]:
import ipywidgets as widgets
year_selector = widgets.IntSlider(
    value=2007,
    min=df["year"].min(),
    max=df["year"].max(),
    step=5,
    description='Year:',
    continuous_update=False
)

@widgets.interact(
    year=year_selector
)
def update_plot(year):
    df_year = df[df["year"] == year].sort_values("pop", ascending=False).head(10)
    fig = px.bar(
        df_year,
        x="country",
        y="pop",
        color="continent",
        title=f"Top 10 Countries by Population in {year}",
        labels={"pop": "Population", "country": "Country"}
    )
    fig.update_xaxes(categoryorder="array", categoryarray=df_year["country"].tolist())
    fig.update_layout(autosize=False, width=800, height=500)
    fig.show()

interactive(children=(IntSlider(value=2007, continuous_update=False, description='Year:', max=2007, min=1952, …

In [68]:
@widgets.interact(
    year=year_selector
)
def update_plot(year):
    df_year = df[df["year"] == year]
    fig = px.scatter(df_year,
                     x="gdpPercap",
                     y="lifeExp",
                     size="pop", color="country",
                     hover_name="country",
                     title=f"GDP per Capita vs Life Expectancy - {year}",
                     labels={"gdpPercap": "GDP per Capita", "lifeExp": "Life Expectancy"},
    )
    fig.update_layout(autosize=False, width=800, height=500)
    fig.show()


interactive(children=(IntSlider(value=1967, continuous_update=False, description='Year:', max=2007, min=1952, …

<a name="demonstration"></a>
## Demonstration

### 1. Introduction

This demonstration summarizes the key insights from the exploratory analysis of the Gapminder dataset.
It shows how life expectancy, economic development, and population have changed across countries and continents.

### 2. Country-Level Trends: Improvements in Life Expectancy

Countries show different patterns in how life expectancy changes over time.
**Japan** increases steadily from the 1950s as healthcare improves.
**Zambia** and **Zimbabwe** show declines in the 1990s due to health crises, followed by slow recovery.
**Oman** shows one of the largest improvements, rising from very low levels to near high‑income countries.

These country-level differences motivate a broader look at regional patterns.

To illustrate how life expectancy has evolved at the country level,
I provide an interactive line chart that allows the user to select any country.
This makes it possible to explore differences in development pace and historical context.
For example, selecting Japan shows a rapid increase in life expectancy after the 1950s.

In [31]:
import ipywidgets as widgets
from ipywidgets import interact
import plotly.express as px

country_options = {
    f"{row['continent']} - {row['country']}": row['country']
    for _, row in df[['continent','country']].drop_duplicates().iterrows()
}
country_options = dict(sorted(country_options.items(), key=lambda item: item[0]))

# Create a dropdown with Japan as the default value
country_selector = widgets.Dropdown(
    options=country_options,
    value="Japan",
    description='Country:',
    style={'description_width': 'initial'}
)

@interact(country=country_selector)
def plot_country_life_expectancy(country):
    df_country = df[df['country'] == country]
    fig = px.line(
        df_country,
        x="year",
        y="lifeExp",
        title=f"Life Expectancy Over Time: {country}",
        labels={"lifeExp": "Life Expectancy", "year": "Year"}
    )
    fig.update_layout(autosize=False)
    fig.update_yaxes(
        range=[
            df["lifeExp"].min() * 0.95,
            df["lifeExp"].max() * 1.05
        ]
    )
    fig.show()


interactive(children=(Dropdown(description='Country:', index=88, options={'Africa - Algeria': 'Algeria', 'Afri…

These country‑level patterns motivate a broader question:
how do these trends differ across continents?  

### 3. Continental Patterns: Regional Differences in Development

The previous section highlights how individual countries have changed over time.
Examining broader regional patterns helps explain why these trends differ across the world.
Continents vary widely in economic development, demographic growth, and public‑health conditions,
and these structural differences shape life expectancy outcomes.

Aggregating the data by continent reveals clear regional differences.

- **Europe** and **Oceania** maintain high life expectancy throughout the period.
- **Asia** improves rapidly after 1970.
- **Africa** increases slowly and remains lower than other regions.

GDP per capita shows a similar pattern: regions with faster economic growth tend to improve life expectancy more quickly. Population trends also show that Asia and Africa drive most global population growth.

(Continent-level GDP per capita is calculated as the total GDP of a continent divided by its total population. Continent-level life expectancy is calculated as the simple average of life expectancies across countries within that continent. This approach is used because the goal is to compare countries equally and examine regional inequalities. Using a weighted average would reduce the influence of small-population countries and make these differences harder to see.)

In [32]:
# 1. Average Life Expectancy by Continent
fig = px.line(
    df_continent,
    x="year",
    y="lifeExp_mean",
    color="continent",
    title="Average Life Expectancy by Continent",
    labels={"lifeExp_mean": "Average Life Expectancy", "year": "Year"}
)
fig.update_layout(width=900, height=500)
fig.show()

# 2. Average GDP per Capita by Continent
fig = px.line(
    df_continent,
    x="year",
    y="gdp_percap",
    color="continent",
    title="Average GDP per Capita by Continent",
    labels={"gdp_percap": "Average GDP per Capita", "year": "Year"}
)
fig.update_layout(width=900, height=500)
fig.show()

# 3. Total Population by Continent
fig = px.line(
    df_continent,
    x="year",
    y="pop_sum",
    color="continent",
    title="Total Population by Continent",
    labels={"pop_sum": "Total Population", "year": "Year"}
)
fig.update_layout(width=900, height=500)
fig.show()


### 4. Global Inequality: Are Countries Becoming More Similar Over Time?

The continental differences observed in Section 3 show that regions follow distinct development patterns.
These differences help explain part of the inequality seen within and across regions.

To assess global inequality directly, we examine the standard deviation of life expectancy across all countries over time.


In [ ]:
df_std_lifeExp = df.groupby("year")["lifeExp"].std().reset_index()

fig = px.line(
    df_std_lifeExp,
    x="year",
    y="lifeExp",
    title="Global Life Expectancy Inequality Over Time (Standard Deviation)",
    labels={"lifeExp": "Life Expectancy Standard Deviation", "year": "Year"}
)
fig.update_layout(width=900, height=500)
fig.show()


The global standard deviation does not follow a simple downward trend.
It fell from the 1950s to the mid‑1980s, rose again through the 1990s and early 2000s, and then became slightly more stable in recent years.
This means countries sometimes become more similar and sometimes more different.

The standard deviation alone does not tell us which regions contribute most to global inequality.
However, the trends in Section 3 suggest that slower improvements in parts of Africa help keep these gaps in place.

GDP per capita shows a similar pattern of regional differences.
This suggests that economic conditions and health outcomes are related,
although their global inequality patterns do not always move in the same direction.

In [ ]:
df_std_gdp = df.groupby("year")["gdpPercap"].std().reset_index()

fig = px.line(
    df_std_gdp,
    x="year",
    y="gdpPercap",
    title="Global GDP per Capita Inequality Over Time (Standard Deviation)",
    labels={"gdpPercap": "GDP per Capita Standard Deviation", "year": "Year"}
)
fig.update_layout(width=900, height=500)
fig.show()


This indicates that economic inequality, like health inequality, has experienced periods of convergence and divergence rather than a steady decline.

Together, these findings indicate that although global health has improved overall,
but the pace of improvement still varies widely across regions.

### 5. Relationships Between Indicators: What Drives Life Expectancy?

To understand the drivers behind the inequalities shown in Section 4,
this section analyzes the relationships among GDP per capita, life expectancy, and population.

Scatter plots show the relationships between key variables:

- **GDP per capita vs. Life Expectancy** shows a strong positive association.
  Wealthier countries tend to have longer life expectancy due to better healthcare systems, infrastructure, and living conditions.
- **Population vs. GDP per capita** shows almost no meaningful relationship.
  Large populations do not imply higher or lower economic development.
- **Population vs. Life Expectancy** also shows weak correlation.
  Population size is influenced by fertility rates, age structure, and migration rather than health outcomes.


In [41]:
fig = px.scatter(
    df,
    x="gdpPercap",
    y="lifeExp",
    color="continent",
    hover_name="country",
    title=f"GDP per Capita vs Life Expectancy (correlation: {df['gdpPercap'].corr(df['lifeExp']):.2f})",
    labels={"gdpPercap": "GDP per Capita", "lifeExp": "Life Expectancy"},
    opacity=0.5,
    log_x=True
)
fig.update_layout(width=900, height=500)
fig.show()


In [42]:
fig = px.scatter(
    df,
    x="pop",
    y="gdpPercap",
    color="continent",
    hover_name="country",
    title=f"Population vs GDP per Capita (correlation: {df['pop'].corr(df['gdpPercap']):.2f})",
    labels={"pop": "Population", "gdpPercap": "GDP per Capita"},
    opacity=0.5,
    log_x=True,
    log_y=True
)
fig.update_layout(width=900, height=500)
fig.show()


In [43]:
fig = px.scatter(
    df,
    x="pop",
    y="lifeExp",
    color="continent",
    hover_name="country",
    title=f"Population vs Life Expectancy (correlation: {df['pop'].corr(df['lifeExp']):.2f})",
    labels={"pop": "Population", "lifeExp": "Life Expectancy"},
    opacity=0.5,
    log_x=True
)
fig.update_layout(width=900, height=500)
fig.show()


These results confirm that economic development, not population size, is the primary driver of improvements in life expectancy.

To better understand the variability within each continent, three additional complementary visualizations are used:

- **jittered points** show each individual country,
- **box plots** summarize the central tendency and interquartile range, and
- **violin plots** illustrate the full distribution shape.

Together, these three views provide a complete picture of both the spread and structure of life expectancy within each region.
In other words, continent-wise disparities.

In [38]:
fig = px.violin(
    df,
    x="continent",
    y="lifeExp",
    color="continent",
    box=True,
    points="all",
    title="Life Expectancy Distribution by Continent",
    labels={"continent": "Continent", "lifeExp": "Life Expectancy"}
)
fig.update_layout(width=900, height=500)
fig.show()


### 6. Interactive Dashboard: Bringing the Analysis Together

Here, you can explore the country-level trends and characteristics based on the knowledge gained from the previous sections.
The dashboard allows users to explore long‑term trends, global standing, and economic context for any selected country and year.
It consists of three linked visualizations:

- the time‑series chart shows the country’s long‑term trend, with the selected year highlighted.  
- the histogram places that year’s life expectancy within the global distribution using fixed five‑year bins. 
- the scatter plot relates life expectancy to GDP per capita, helping explain why the country falls into its observed position. 

Together, these visualizations offer a coherent narrative of historical change, global standing, and economic context.

In [ ]:
import ipywidgets as widgets
from ipywidgets import interact
import plotly.express as px
import numpy as np
from IPython.display import Markdown
from IPython.display import HTML
from plotly.subplots import make_subplots


# Widgets
country_selector = widgets.Dropdown(
    options=country_options,
    value='Japan',
    description='Country:',
    style={'description_width': 'initial'}
)

year_selector = widgets.IntSlider(
    value=df['year'].min(),
    min=df['year'].min(),
    max=df['year'].max(),
    step=5,
    description='Year:',
    continuous_update=False,
    style={'description_width': 'initial'}
)

chart_margin = dict(t=40, b=40, l=40, r=40)
title_x = 0.5                 # center title
title_y = 0.95                # adjust vertical position of title
title_pad = {'t': 0, 'b': 0}  # adjust margin between title and graph
width = 900
horizontal_line = HTML("<hr style='border: 0; border-top: 1px solid #ccc; margin: 10px 0; width: " + str(width) + "px;'>")

# Dashboard
# @interact(country=country_selector, year=year_selector)
def dashboard(country, year):

    df_country = df[df['country'] == country]
    df_year = df[df['year'] == year]

    # 1. Life Expectancy Over Time
    fig1 = px.line(
        df_country,
        x="year",
        y="lifeExp",
        labels={"lifeExp": "Life Expectancy", "year": "Year"}
    )
    y_min = (df["lifeExp"].min() // 5) * 5
    y_max = ((df["lifeExp"].max() // 5) + 1) * 5

    fig1.update_yaxes(range=[y_min, y_max])

    # vertical line for the selected year
    fig1.add_vline(
        x=year,
        line_width=2,
        line_dash="dash",
        line_color="red"
    )

    fig1.update_layout(
        margin=chart_margin,
        width=width, height=350,
        title=dict(
            text=f"Life Expectancy Over Time: {country} - Highlighting {year}",
            x=title_x,
            y=title_y,
            pad=title_pad
        )
    )
    # fig1.show()
    display(horizontal_line)
    # df_year = df[df['year'] == year]

    # 
    global_min = df["lifeExp"].min()
    global_max = df["lifeExp"].max()
    
    bin_start = (global_min // 5) * 5
    bin_end   = ((global_max // 5) + 1) * 5

    bins = np.arange(bin_start, bin_end + 1, 5)

    bin_centers = (bins[:-1] + bins[1:]) / 2

    selected_life = df_year[df_year['country'] == country]["lifeExp"].iloc[0]

    selected_bin_index = np.digitize(selected_life, bins) - 1

    counts = []
    for i in range(len(bins) - 1):
        bin_min = bins[i]
        bin_max = bins[i+1]
        count = ((df_year["lifeExp"] >= bin_min) & (df_year["lifeExp"] < bin_max)).sum()
        counts.append(count)

    colors = ["lightblue"] * len(counts)
    if 0 <= selected_bin_index < len(colors):
        colors[selected_bin_index] = "red"

    fig2 = go.Figure()

    fig2.add_bar(
        x=bin_centers,
        y=counts,
        marker_color=colors,
        width=5,
        marker_line_width=0,
        name="Life Expectancy Distribution",
        hovertext=[f"Life Expectancy range: {math.floor(bin_centers[i] - 2.5)} to {math.ceil(bin_centers[i] + 2.5)}<br>Count: {counts[i]}" for i in range(len(bin_centers))]
    )
    fig2.update_traces(
        hovertemplate="<b>%{hovertext}</b><extra></extra>"
    )

    fig2.update_layout(
        margin=chart_margin,
        title=dict(
            text=f"Life Expectancy Distribution ({year}) - Highlighting {country} bin",
            x=title_x,
            y=title_y,
            pad=title_pad
        ),
        xaxis_title="Life Expectancy",
        yaxis_title="Count",
        bargap=0,
        width=width,
        height=350
    )

    # fig2.show()
    # fig1_html = fig1.to_html(include_plotlyjs=False, full_html=False)
    # fig2_html = fig2.to_html(include_plotlyjs=False, full_html=False)
    # display(horizontal_line)
    # box = widgets.HBox([
    #     widgets.HTML(fig1_html),
    #     widgets.HTML(fig2_html)
    # ])
    # display(box)
    fig = make_subplots(
        rows=1,
        cols=2,
        # column_titles=("Life Expectancy Over Time", "Life Expectancy Distribution"),
        subplot_titles=(
            f"Life Expectancy Over Time ({country})<br />Highlighting {year}",
            f"Life Expectancy Distribution ({year})<br />Highlighting {country} bin"
        ),
        # row_titles=("Life Expectancy"),
        # y_title=("Life Expectancy", "Count"),
    )

    fig.add_trace(fig1.data[0], row=1, col=1)
    fig.add_trace(fig2.data[0], row=1, col=2)
    fig.update_xaxes(title_text="Year", row=1, col=1)
    fig.update_yaxes(
        title_text="Life Expectancy",
        range=[y_min, y_max], row=1, col=1,
        ticklabelstandoff=8
    )
    fig.update_xaxes(title_text="Life Expectancy", row=1, col=2)
    fig.update_yaxes(
        title_text="Count",
        ticklabelstandoff=8,
        row=1, col=2)
    # remove legend for the second subplot
    fig.update_layout(showlegend=False)
    # add vertical line for the selected year in the first subplot
    fig.add_vline(
        x=year,
        line_width=2,
        line_dash="dash",
        line_color="red",
        row=1, col=1
    )


    fig.update_layout(width=width, height=350)
    fig.show()

    # html = """
    # <div style="display: flex; justify-content: space-between; align-items: flex-start; gap: 10px;">
    #     <div style="flex: 1;">
    #         {fig1}
    #     </div>
    #     <div style="flex: 1;">
    #         {fig2}
    #     </div>
    # </div>
    # """

    # display(HTML(html.format(
    #     fig1=fig1.to_html(include_plotlyjs=False, full_html=False),
    #     fig2=fig2.to_html(include_plotlyjs=False, full_html=False)
    # )))
    # ---

    fig3 = px.scatter(
        df_year[df_year['country'] != country],
        x="gdpPercap",
        y="lifeExp",
        color="continent",
        hover_name="country",
        log_x=True,
        opacity=0.5,
        labels={"gdpPercap": "GDP per Capita (log scale)", "lifeExp": "Life Expectancy"}
    )

    fig3.update_traces(
        hovertemplate="<b>%{hovertext}</b><br>GDP: %{x}<br>LifeExp: %{y}<extra></extra>"
    )
    fig3.update_yaxes(
        # title_text="Life Expectancy",
        ticklabelstandoff=8,
        range=[y_min, y_max]
    )
    fig3.update_xaxes(
        range=[math.log10(df["gdpPercap"].min() * 0.9), math.log10(df["gdpPercap"].max() * 1.1)],
    )

    df_selected = df_year[df_year['country'] == country]

    fig3.add_scatter(
        x=df_selected["gdpPercap"],
        y=df_selected["lifeExp"],
        mode="markers",
        marker=dict(size=18, color="red", symbol="star"),
        hovertext=df_selected["country"],
        hovertemplate="<b>%{hovertext} (selected)</b><br>GDP: %{x}<br>LifeExp: %{y}<extra></extra>",
        name=f"{country} (selected)"
    )

    fig3.update_layout(
        width=width, height=350,
        margin=chart_margin,
        title=dict(
            text=f"GDP per Capita vs Life Expectancy ({year}) - Highlighting {country}",
            x=title_x,
            y=title_y,
            pad=title_pad
        )
    )
    fig3.show()

widgets.interact(
    dashboard,
    country=country_selector,
    year=year_selector
)

interactive(children=(Dropdown(description='Country:', index=88, options={'Africa - Algeria': 'Algeria', 'Afri…

<function __main__.dashboard(country, year)>

### 7. Conclusion

The demonstration highlights several key insights:

- Life expectancy improvements are closely related to economic development.
- Regional differences remain, and global health inequality has changed over time rather than decreasing steadily.
- Population size is largely independent of both GDP per capita and life expectancy.
- Continental trends reveal distinct development trajectories. They shape global patterns.

Together with the interactive dashboard, these findings give a clear and accessible view of how countries and continents have changed in health, population, and economic development.